---
title: "Statistical Analysis with Sex as a Factor for: Matrisome alterations in obesity - Adipose tissue transcriptome study on monozygotic  weight-discordant twins"
output: html_notebook
---

### Library Used
```{r results='hide', eval=FALSE}
#General Bioconductor packages
library(Biobase)
library(oligoClasses)

#Annotation and data import packages
library(ArrayExpress)
library(pd.hg.u133.plus.2) 
library(biomaRt)

#Quality control and pre-processing packages
library(oligo)
library(arrayQualityMetrics)

#Analysis and statistics packages
library(limma)
library(statmod)

#Plotting and color options packages
library(gplots)
library(ggplot2)
library(RColorBrewer)

#Formatting/documentation packages
library(dplyr)
library(tidyr)

#Helpers:
library(openxlsx)
library(readxl, writexl)
```

# Data Preparation
In order to follow the pipeline for the analysis, the genomic data must be formatted to fit into an ExpressionSet class from Bioconductor. The class includes and assay data, a feature data, and a phenotypic meta-data. Assay data and phenotype data come from the Finnish project while the feature data is generated with the biomaRT library. In addition to the group lean vs heavy, other factors such as weight, bmi, height and sex was included as supplemental data for the phynotype data.

There are 3 pairs of twins in the adipose tissue data whose supplemental information was not provided. For the sake of analysis including the supplemental data i.e. sex, the three families were removed from the data analysis.
```{r results='hide', eval=FALSE}
wdir <- dirname(rstudioapi::getSourceEditorContext()$path)

### Load adipose tissue gene
adipose_tissue_gene_full <- read_excel(paste(wdir, "/Datasets/MZtwins_whole genome_adipocyte_adipose tissue.xlsx", sep = ""), 
                                  sheet = "Adipose Tissue", skip = 4, col_names = FALSE)

## Remove the 3 pairs of twins without supplemental data
adipose_tissue_gene <- adipose_tissue_gene_full[-c(6,8,14,31,33,39)]

### Load supplemental data about each individual
adipose_tissue_supplementalData_lean <- read_excel(paste(wdir, "/Datasets/Supplementary Table 1_Twin characteristics_adipocity_July 15_2020.xlsx", sep = ""),
                                                   sheet = "22 Twins", range = "C4:F25", col_names = FALSE)
adipose_tissue_supplementalData_heavy <- read_excel(paste(wdir, "/Datasets/Supplementary Table 1_Twin characteristics_adipocity_July 15_2020.xlsx", sep = ""),
                                                    sheet = "22 Twins", range = "C32:F53", 
                                                    col_names = FALSE)

```

## Supplemental Data
The order of the family and individual ID's correspond to the order in the genomic data.
```{r results='hide'}
# Family ID
fid_wParam <- paste0('fam', c('Q','S','R','P','U','T','V','O','X','W','N','Y','J','H','E','G','C','K','M','A','B','I',
                              'Q','S','R','P','U','T','V','O','X','W','N','Y','J','H','E','G','C','K','M','A','B','I'))
fid_wParam <- as.factor(fid_wParam)

# Individual ID
pid_wParam <- paste0('ind', sprintf('%02d' , c(17,19,18,16,21,20,22,15,24,23,14,25,10,8,5,7,3,11,13,1,2,9,
                                               42,44,43,41,46,45,47,40,49,48,39,50,35,33,30,32,28,36,38,26,27,34)))
pid_wParam <- as.factor(pid_wParam)

# Create the supplemental data
adipose_tissue_supplementalData <- data.frame(fid_wParam, pid_wParam, rbind(adipose_tissue_supplementalData_lean, adipose_tissue_supplementalData_heavy))

supplemental_header <- c("sex", "height", "weight", "bmi")
colnames(adipose_tissue_supplementalData) <- c("fid", "pid", supplemental_header)

adipose_tissue_supplementalData$sex[adipose_tissue_supplementalData$sex == 1] <- "male"
adipose_tissue_supplementalData$sex[adipose_tissue_supplementalData$sex == 2] <- "female"

adipose_tissue_supplementalData <- adipose_tissue_supplementalData[with(adipose_tissue_supplementalData, order(pid)),]
```

## Phenotype Data
```{r results='hide'}
# Family ID
fid <- paste0('fam', c('A','B','C','E','G','H','I','J','K','M','N','O','P','Q','R','S','T','U','V','W','X','Y',
                       'A','B','C','E','G','H','I','J','K','M','N','O','P','Q','R','S','T','U','V','W','X','Y'))
fid <- as.factor(fid)

# Individual ID
pid <- paste0('ind', sprintf('%02d' , c(1,2,3,5,7,8,9,10,11,13,14,15,16,17,18,19,20,21,22,23,24,25,
                                        26,27,28,30,32,33,34,35,36,38,39,40,41,42,43,44,45,46,47,48,49,50)))
pid <- as.factor(pid)

# Group (lean and heavy)
group <- as.factor(rep(c('lean','heavy'), each=length(pid)/2))

# Create phenotype data
adipose_tissue_phenoData <- data.frame(fid, pid, group, adipose_tissue_supplementalData[,-c(1,2)])

write.csv(adipose_tissue_phenoData, file = paste(wdir, "/Datasets/adipose_tissue_phenoData_suppl.csv", sep=""), 
          row.names = FALSE)
```

## Assay Data
```{r results='hide'}
# Create assay data
adipose_tissue_assayData <- as.data.frame(adipose_tissue_gene[, -c(1, 2)])

rownames(adipose_tissue_assayData) <- make.names(adipose_tissue_gene[[1]], unique=TRUE)
colnames(adipose_tissue_assayData) <- paste(adipose_tissue_phenoData$fid, adipose_tissue_phenoData$pid, sep=".")

write.csv(adipose_tissue_assayData, file = paste(wdir, "/Datasets/adipose_tissue_assayData_suppl.csv", sep=""), 
          row.names = TRUE)
```

## Feature Data
```{r results='hide'}
### Try to get the annotation data
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl") 
annotations <- getBM(
  attributes=c(
    'hgnc_symbol',
    "ensembl_gene_id",
    'description',
    "external_gene_name",
    'gene_biotype',
    'chromosome_name',
    "affy_hg_u133_plus_2"
  ),
  filters = 'hgnc_symbol', values = rownames(adipose_tissue_assayData), mart = mart, uniqueRows=TRUE)

annotations <- annotations[match(rownames(adipose_tissue_assayData), annotations$hgnc_symbol), ]


### Get the additional information for each individual
adipose_tissue_featureData <- data.frame(geneId = as.factor(adipose_tissue_gene[[1]]), 
                                        geneDesc = as.factor(adipose_tissue_gene[[2]]))

write.csv(annotations, file = paste(wdir, "/Datasets/adipose_tissue_featureData_suppl.csv", sep=""), 
          row.names = FALSE)
```

## Create ExpressionSet
```{r}
assayData <- read.csv(paste(wdir, "/Datasets/adipose_tissue_assayData_suppl.csv", sep=""), row.names=1)
assayData <- as.matrix(assayData)

adipose_tissue_phenoData <- read.csv(paste(wdir, "/Datasets/adipose_tissue_phenoData_suppl.csv", sep=""))
rownames(adipose_tissue_phenoData) = colnames(assayData)

metadata <- data.frame(labelDescription=
                         c("Family Id",
                           "Individual Id",
                           "Case/control status",
                           "1=male, 2=female",
                           "cm",
                           "kg",
                           "bmi"),
                       row.names=c("fid", "pid", "group", "sex", "height", "weight", "bmi"))

phenoData <- new("AnnotatedDataFrame", data=adipose_tissue_phenoData, varMetadata=metadata)
adipose_tissue_featureData <- read.csv(paste(wdir, "/Datasets/adipose_tissue_featureData_suppl.csv", sep=""))
rownames(adipose_tissue_featureData) = rownames(assayData)
featureData <- new("AnnotatedDataFrame", data=adipose_tissue_featureData)

raw_data  <- new("ExpressionSet", exprs = assayData, phenoData = phenoData,
                 featureData = featureData,
                 annotation="pd.hg.u133.plus.2")
```

# Data Preprocessing
### Normalization
The data was normalized and filtered. Normalization of the data was performed using the corresponding function from the Oligo library. 
```{r}
oligo::boxplot(raw_data, target = "core", 
               main = "AT: Boxplot of log2-intensitites for the raw data")

palmieri_eset_norm <- oligo::normalize(raw_data)

oligo::boxplot(palmieri_eset_norm, target = "core", 
               main = "AT: Boxplot of log2-intensitites for the normalized data")
```

### Filtering
The data was filtered using a threshold value of 3. Transcripts that do not have intensities larger than the threshold in at least as many arrays as the smallest experimental group were excluded.
```{r}
par(mfrow=c(2,1))

palmieri_medians <- rowMedians(Biobase::exprs(palmieri_eset_norm))

man_threshold <- 3

hist_res <- hist(palmieri_medians, 100, col = "cornsilk", freq = FALSE, 
                 main = "AT: Histogram of the median intensities of unfiltered expression data",
                 border = "antiquewhite4",
                 xlab = "Median intensities")

abline(v = man_threshold, col = "coral4", lwd = 2)

no_of_samples <- table(pData(palmieri_eset_norm)$group)
samples_cutoff <- min(no_of_samples)

idx_man_threshold <- apply(Biobase::exprs(palmieri_eset_norm), 1,
                           function(x){
                             sum(x > man_threshold) >= samples_cutoff})

palmieri_manfiltered <- subset(palmieri_eset_norm, idx_man_threshold)

hist(rowMedians(Biobase::exprs(palmieri_manfiltered)), 100, col = "cornsilk", freq = FALSE, 
     main = "AT: Histogram of the median intensities of filtered expression data",
     border = "antiquewhite4",
     xlab = "Median intensities")
```

# Hypothesis Testing
Regression analysis was performed, and hypothesis testing followed. 

### Design Matrix
```{r}
fam <- as.character(Biobase::pData(palmieri_manfiltered)$fid)
group <- Biobase::pData(palmieri_manfiltered)$group
sex<- Biobase::pData(palmieri_manfiltered)$sex
height <- Biobase::pData(palmieri_manfiltered)$height
weight <- Biobase::pData(palmieri_manfiltered)$weight
bmi <- Biobase::pData(palmieri_manfiltered)$bmi

design <- model.matrix(~ 0 + group)
rownames(design) <- Biobase::pData(palmieri_manfiltered)$pid
```

### Contrast Matrix
```{r}
contrast_matrix <- makeContrasts(groupheavy - grouplean, levels = design)
```

### Model Fit
*voom* function adjusts the data for heteroscedasticity and prepares the data for linear fit. The design and contrast matrices were fit in a linear model using the corresponding function from the limma library. Empirical Bayes method was used to make more precise estimates of gene-wise variability and to get moderated results.
```{r}
par(mfrow=c(1,2))

v <- voom(palmieri_manfiltered, design, plot=TRUE)

dupfit <- duplicateCorrelation(v, design, block = fam)
vfit <- lmFit(v, design, correlation=dupfit$consensus, block = fam)
cfit <- contrasts.fit(vfit, contrasts=contrast_matrix)
efit <- eBayes(cfit)

contrastfit <- contrasts.fit(vfit, contrasts=contrast_matrix)
efit <- eBayes(contrastfit)

plotSA(efit, main="AT: Final model: Mean-variance trend")
```

# Results
Significance of the differential expression of each genes was extracted using the function topTable on the linear model fit. The adjusted p-value is derived from the Benjamini-Hochberg Procedure, which decreases the false discovery rate.
```{r}
table_adipose <- topTable(efit, number = Inf, sort.by = "p")

dt <- decideTests(efit)
summary(dt)

hist(table_adipose$adj.P.Val, main = "AT: Heavy vs Lean P-values intensities", xlab = "p-values")
```
Significance of differential gene expressions was determined based on the p-values calculated by the Benjamini-Hochberg procedure.
```{r}
signif_genes <- rownames(as.matrix(dt[dt != 0,]))
signif_toptable <- table_adipose[rownames(table_adipose) %in% signif_genes,]

adjPval_min <- min(table_adipose$adj.P.Val)
adjPval_max <- max(table_adipose$adj.P.Val)

adjPval_cutoff <- max(signif_toptable$adj.P.Val)
```

# Matrisome Analysis
The data analysis was performed on the entire genomic dataset. The results were then filtered for the matrisome genes only. Matrisome genes were taken from the Matrisome Project database.
```{r}
matrisome_masterlist <- read_excel(paste(wdir, "/Public Datasets/matrisome_hs_masterlist.xls", sep = ""))

ECM_glycoproteins <- matrisome_masterlist[matrisome_masterlist$Category == "ECM Glycoproteins",]
collagens <- matrisome_masterlist[matrisome_masterlist$Category == "Collagens",]
proteoglycans <- matrisome_masterlist[matrisome_masterlist$Category == "Proteoglycans",]
ECM_affiliated_proteins <- matrisome_masterlist[matrisome_masterlist$Category == "ECM-affiliated Proteins",]
ECM_regulators <- matrisome_masterlist[matrisome_masterlist$Category == "ECM Regulators",]
secreted_factors <- matrisome_masterlist[matrisome_masterlist$Category == "Secreted Factors",]
```

### Significant Matrisome Genes
```{r}
matrisome_toptable <- table_adipose[rownames(table_adipose) %in% matrisome_masterlist$`Gene Symbol`,]

matrisome_toptable$`Differential Expression` <- "Not Sig."
matrisome_toptable$`Differential Expression`[matrisome_toptable$logFC < 0 & -log10(matrisome_toptable$adj.P.Val) > -log10(adjPval_cutoff)] <- "Matrisome Down"
matrisome_toptable$`Differential Expression`[matrisome_toptable$logFC > 0 & -log10(matrisome_toptable$adj.P.Val) > -log10(adjPval_cutoff)] <- "Matrisome Up"
matrisome_toptable$label <- NA
matrisome_toptable$label[matrisome_toptable$`Differential Expression` %in% c("Matrisome Down", "Matrisome Up")] <- rownames(matrisome_toptable[matrisome_toptable$`Differential Expression` %in% c("Matrisome Down", "Matrisome Up"),])

matr_signif_genes <- signif_genes[signif_genes %in% matrisome_masterlist$`Gene Symbol`]

matrisome_dt <- as.matrix(dt[rownames(dt) %in% matrisome_masterlist$`Gene Symbol`,])
dt_genes <- rownames(matrisome_dt)
summary(matrisome_dt)
```

# Plots
```{r}
table_adipose$`Differential Expression` <- "Not Sig."
table_adipose$`Differential Expression`[table_adipose$logFC < 0 & -log10(table_adipose$adj.P.Val) > -log10(adjPval_cutoff)] <- "Transcriptome Down"
table_adipose$`Differential Expression`[table_adipose$logFC > 0 & -log10(table_adipose$adj.P.Val) > -log10(adjPval_cutoff)] <- "Transcriptome Up"
table_adipose$`Differential Expression`[table_adipose$`Differential Expression` == "Transcriptome Down" & 
                                          table_adipose$ensembl_gene_id %in% downregulated_matr_table$ensembl_gene_id] <- "Matrisome Down"
table_adipose$`Differential Expression`[table_adipose$`Differential Expression` == "Transcriptome Up" & 
                                          table_adipose$ensembl_gene_id %in% upregulated_matr_table$ensembl_gene_id] <- "Matrisome Up"

matrisome_toptable$`Differential Expression` <- "Not Sig."
matrisome_toptable$`Differential Expression`[matrisome_toptable$logFC < 0 & -log10(matrisome_toptable$adj.P.Val) > -log10(adjPval_cutoff)] <- "Matrisome Down"
matrisome_toptable$`Differential Expression`[matrisome_toptable$logFC > 0 & -log10(matrisome_toptable$adj.P.Val) > -log10(adjPval_cutoff)] <- "Matrisome Up"

# Label the significant matrisome genes only
table_adipose$label <- NA
table_adipose$label[table_adipose$`Differential Expression` %in% c("Matrisome Down", "Matrisome Up")] <- rownames(table_adipose[table_adipose$`Differential Expression` %in% c("Matrisome Up", "Matrisome Down"),])

```

##Volcano Plots
### Transcriptome data
```{r}
volccol <- c("grey60", "grey60", "blue", "darkorange",  "grey87")
names(volccol) <- c("Transcriptome Down", "Transcriptome Up", "Matrisome Down", "Matrisome Up", "Not Sig.")

ggplot(data = table_adipose, aes(x=logFC, y=-log10(adj.P.Val), label=label, col=`Differential Expression`)) + 
  scale_x_continuous(labels = c(-0.50, -0.25, 0, 0.25, 0.50)) +
  scale_y_continuous(breaks = c(0, 1, 2, 3, 4, 5, 6), labels = c(0, 1, 2, 3, 4, 5, 6), limits = c(0, 6)) +
  geom_point(size=3) + 
  geom_point(size=3, data=subset(table_adipose, `Differential Expression` %in% c("Matrisome Down", "Matrisome Up"))) +
  geom_vline(xintercept=c(0, 0), col="black", linetype="dashed", size = 0.75) +
  geom_hline(yintercept=c(-log10(adjPval_cutoff)), col="black", linetype="dashed", size = 0.75) +
  scale_colour_manual(values = volccol) + 
  # geom_text(nudge_y = -0.15, check_overlap = TRUE, colour = "black") +
  # geom_text_repel(nudge_y = -0.1) +
  theme(plot.title = element_text(hjust = 0.5),
        text = element_text(size = 8),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black", size = 0.75),
        axis.title = element_blank(),
        axis.text = element_text(size = 20, family = 'Helvetica'),
        # legend.key = element_rect(fill = "white"),
        # legend.title = element_blank(),
        legend.position = "none", 
        plot.margin = unit(c(1, 1, 1, 1), "cm"))
```
### Matrisome Data
```{r}
ggplot(data = matrisome_toptable, aes(x=logFC, y=-log10(adj.P.Val), label=label, col=`Differential Expression`)) + 
  scale_x_continuous(labels = c(-0.50, -0.25, 0, 0.25, 0.50)) + 
  scale_y_continuous(breaks = c(0, 1, 2, 3, 4, 5), labels = c(0, 1, 2, 3, 4, 5), limits = c(0, 5)) +
  geom_point(size=3) + 
  geom_vline(xintercept=c(0, 0), col="black", linetype="dashed", size=0.75) +
  geom_hline(yintercept=c(-log10(adjPval_cutoff)), col="black", linetype="dashed", size=0.75) +
  scale_colour_manual(values = volccol) + 
  # geom_text(colour = "black", size = 4, position = position_nudge(y = 0.1), check_overlap = TRUE) +
  geom_text_repel(colour = "black", size = 4, box.padding = 0.25) +
  theme(plot.title = element_text(hjust = 0.5),
        text = element_text(size = 8),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black", size = 0.75),
        axis.title = element_blank(),
        axis.text = element_text(size = 20, family = 'Helvetica'),
        # legend.text = element_text(family = "Helvetica", size = 12),
        # legend.key = element_blank(),
        # legend.title = element_blank(),
        # legend.position = c(0.15,0.93),
        legend.position = "none",
        plot.margin = unit(c(1, 1, 1, 1), "cm"))
```




### Additional Tests
Additional tests were performed to show the effect of the parameters on the data. This includes a combination of the parameters as factors in the design matrix.

# Interaction effect of sex
```{r}
interVar <- paste(sex, group, sep=".")
interVar <- factor(interVar, levels=c("male.lean","male.heavy","female.lean","female.heavy"))

design1 <- model.matrix(~0 + interVar)
rownames(design1) <- Biobase::pData(palmieri_manfiltered)$pid 

contrast_matrix1 <- makeContrasts(
  HeavyLeanInMale=interVarmale.heavy - interVarmale.lean,
  HeavyLeanInFem=interVarfemale.heavy - interVarfemale.lean,
  Diff=(interVarmale.heavy-interVarmale.lean) - (interVarfemale.heavy-interVarfemale.lean),
  levels=design1)
v1 <- voom(palmieri_manfiltered, design1, plot = FALSE)

dupfit1 <- duplicateCorrelation(v1, design1, block=fam)
vfit1 <- lmFit(v1, design1, correlation=dupfit1$consensus, block = fam)
cfit1 <- contrasts.fit(vfit1, contrast_matrix1)
efit1 <- eBayes(fit1)

plotSA(efit1, main="AT: Final model: Mean-variance trend")

dt1 <- decideTests(efit1, method = "global")
summary(dt1)

vennDiagram(dt1)
```
# Sex as a factor
```{r}
par(mfrow=c(1,2))

design2 <- model.matrix(~0 + group + sex)
rownames(design2) <- Biobase::pData(palmieri_manfiltered)$pid 

contrast_matrix2 <- makeContrasts(groupheavy - grouplean, levels = design2)

v2 <- voom(palmieri_manfiltered, design2, plot=TRUE)

dupfit2 <- duplicateCorrelation(v2, design2, block=fam)
vfit2 <- lmFit(v2, design2, correlation=dupfit2$consensus, block = fam)
cfit2 <- contrasts.fit(vfit2, contrasts=contrast_matrix2)
efit2 <- eBayes(cfit2)

plotSA(efit2, main="AT: Final model: Mean-variance trend")

dt2 <- decideTests(efit2, method = "global")
summary(dt2)
```

# Group + fam as a factor
```{r}
par(mfrow=c(1,2))

design3 <- model.matrix(~ 0 + group + fam)
rownames(design3) <- Biobase::pData(palmieri_manfiltered)$pid

contrast_matrix3 <- makeContrasts(groupheavy - grouplean, levels = design3)

v3 <- voom(palmieri_manfiltered, design3, plot=TRUE)

vfit3 <- lmFit(v3, design3)
cfit3 <- contrasts.fit(vfit3, contrasts=contrast_matrix3)
efit3 <- eBayes(cfit3)

plotSA(efit3, main="AT: Final model: Mean-variance trend")

table_adipose3 <- topTable(efit3, number = Inf, sort.by = "p")

dt3 <- decideTests(efit3)
summary(dt3)
```


